// cdk8s-plus/lib/job.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobSpecDefinition = exports.Job = void 0;
const base_1 = require("./base");
const cdk8s_1 = require("cdk8s");
const k8s = require("./imports/k8s");
const pod_1 = require("./pod");
const utils_1 = require("./utils");
/**
 * A Job creates one or more Pods and ensures that a specified number of them successfully terminate. As pods successfully complete,
 * the Job tracks the successful completions. When a specified number of successful completions is reached, the task (ie, Job) is complete.
 * Deleting a Job will clean up the Pods it created. A simple case is to create one Job object in order to reliably run one Pod to completion.
 * The Job object will start a new Pod if the first Pod fails or is deleted (for example due to a node hardware failure or a node reboot).
 * You can also use a Job to run multiple Pods in parallel.
 */
class Job extends base_1.Resource {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        this.spec = new JobSpecDefinition(props.spec);
        this.apiObject = new k8s.Job(this, 'Default', {
            metadata: props.metadata,
            spec: utils_1.lazy(() => this.spec._toKube()),
        });
    }
}
exports.Job = Job;
class JobSpecDefinition {
    constructor(props = {}) {
        var _a, _b;
        this.podSpecTemplate = new pod_1.PodSpecDefinition({
            restartPolicy: (_b = (_a = props.podSpecTemplate) === null || _a === void 0 ? void 0 : _a.restartPolicy) !== null && _b !== void 0 ? _b : pod_1.RestartPolicy.NEVER,
            ...props.podSpecTemplate,
        });
        this.podMetadataTemplate = new cdk8s_1.ApiObjectMetadataDefinition(props.podMetadataTemplate);
        this.ttlAfterFinished = props.ttlAfterFinished;
    }
    /**
     * @internal
     */
    _toKube() {
        return {
            template: {
                metadata: this.podMetadataTemplate.toJson(),
                spec: this.podSpecTemplate._toKube(),
            },
            ttlSecondsAfterFinished: this.ttlAfterFinished ? this.ttlAfterFinished.toSeconds() : undefined,
        };
    }
}
exports.JobSpecDefinition = JobSpecDefinition;
