// cdk8s-plus/lib/service-account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceAccount = void 0;
const base_1 = require("./base");
const k8s = require("./imports/k8s");
const utils_1 = require("./utils");
/**
 * A service account provides an identity for processes that run in a Pod.
 *
 * When you (a human) access the cluster (for example, using kubectl), you are
 * authenticated by the apiserver as a particular User Account (currently this
 * is usually admin, unless your cluster administrator has customized your
 * cluster). Processes in containers inside pods can also contact the apiserver.
 * When they do, they are authenticated as a particular Service Account (for
 * example, default).
 *
 * @see https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account
 */
class ServiceAccount extends base_1.Resource {
    constructor(scope, id, props = {}) {
        var _a;
        super(scope, id, props);
        this._secrets = (_a = props.secrets) !== null && _a !== void 0 ? _a : [];
        this.apiObject = new k8s.ServiceAccount(this, 'Resource', {
            metadata: props.metadata,
            secrets: utils_1.lazy(() => utils_1.undefinedIfEmpty(this._secrets.map(s => ({ name: s.name })))),
        });
    }
    /**
     * Imports a service account from the cluster as a reference.
     * @param name The name of the service account resource.
     */
    static fromServiceAccountName(name) {
        return { name: name };
    }
    /**
     * Allow a secret to be accessed by pods using this service account.
     * @param secret The secret
     */
    addSecret(secret) {
        this._secrets.push(secret);
    }
    /**
     * List of secrets allowed to be used by pods running using this service
     * account.
     *
     * Returns a copy. To add a secret, use `addSecret()`.
     */
    get secrets() {
        return [...this._secrets];
    }
}
exports.ServiceAccount = ServiceAccount;
