// cdk8s/lib/api-object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiObject = void 0;
const constructs_1 = require("constructs");
const chart_1 = require("./chart");
const _util_1 = require("./_util");
const _tokens_1 = require("./_tokens");
const stringify = require("json-stable-stringify");
const metadata_1 = require("./metadata");
class ApiObject extends constructs_1.Construct {
    /**
     * Defines an API object.
     *
     * @param scope the construct scope
     * @param ns namespace
     * @param options options
     */
    constructor(scope, ns, options) {
        var _a, _b, _c, _d;
        super(scope, ns);
        this.options = options;
        this.chart = chart_1.Chart.of(this);
        this.kind = options.kind;
        this.apiVersion = options.apiVersion;
        this.name = (_b = (_a = options.metadata) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : this.chart.generateObjectName(this);
        this.metadata = new metadata_1.ApiObjectMetadataDefinition({
            name: this.name,
            namespace: (_d = (_c = options.metadata) === null || _c === void 0 ? void 0 : _c.namespace) !== null && _d !== void 0 ? _d : this.chart.namespace,
            // override user defined values
            ...options.metadata,
        });
    }
    /**
     * Create a dependency between this ApiObject and other constructs.
     * These can be other ApiObjects, Charts, or custom.
     *
     * @param dependencies the dependencies to add.
     */
    addDependency(...dependencies) {
        constructs_1.Node.of(this).addDependency(...dependencies);
    }
    /**
     * Renders the object to Kubernetes JSON.
     */
    toJson() {
        const data = {
            ...this.options,
            metadata: this.metadata.toJson(),
        };
        // convert to "pure data" so, for example, when we convert to yaml these
        // references are not converted to anchors.
        return JSON.parse(stringify(_util_1.sanitizeValue(_tokens_1.resolve(this, data))));
    }
}
exports.ApiObject = ApiObject;
